// @microsoft/teams.api/dist/clients/user/token.js
'use strict';

var qs = require('qs');
var http = require('@microsoft/teams.common/http');
var apiClientSettings = require('../api-client-settings');

function _interopDefault (e) { return e && e.__esModule ? e : { default: e }; }

var qs__default = /*#__PURE__*/_interopDefault(qs);

const USER_TOKEN_ENDPOINTS = {
  GET_TOKEN: "api/usertoken/GetToken",
  GET_AAD_TOKENS: "api/usertoken/GetAadTokens",
  GET_STATUS: "api/usertoken/GetTokenStatus",
  SIGN_OUT: "api/usertoken/SignOut",
  EXCHANGE: "api/usertoken/exchange"
};
class UserTokenClient {
  get http() {
    return this._http;
  }
  set http(v) {
    this._http = v;
  }
  _http;
  _apiClientSettings;
  constructor(options, apiClientSettings$1) {
    if (!options) {
      this._http = new http.Client();
    } else if ("request" in options) {
      this._http = options;
    } else {
      this._http = new http.Client(options);
    }
    this._apiClientSettings = apiClientSettings.mergeApiClientSettings(apiClientSettings$1);
  }
  async get(params) {
    const q = qs__default.default.stringify(params);
    const res = await this.http.get(
      `${this._apiClientSettings.oauthUrl}/${USER_TOKEN_ENDPOINTS.GET_TOKEN}?${q}`
    );
    return res.data;
  }
  async getAad(params) {
    const q = qs__default.default.stringify(params);
    const res = await this.http.post(
      `${this._apiClientSettings.oauthUrl}/${USER_TOKEN_ENDPOINTS.GET_AAD_TOKENS}?${q}`,
      params
    );
    return res.data;
  }
  async getStatus(params) {
    const q = qs__default.default.stringify(params);
    const res = await this.http.get(
      `${this._apiClientSettings.oauthUrl}/${USER_TOKEN_ENDPOINTS.GET_STATUS}?${q}`
    );
    return res.data;
  }
  async signOut(params) {
    const q = qs__default.default.stringify(params);
    const res = await this.http.delete(
      `${this._apiClientSettings.oauthUrl}/${USER_TOKEN_ENDPOINTS.SIGN_OUT}?${q}`,
      { data: params }
    );
    return res.data;
  }
  async exchange(params) {
    const q = qs__default.default.stringify({
      userId: params.userId,
      connectionName: params.connectionName,
      channelId: params.channelId
    });
    const res = await this.http.post(
      `${this._apiClientSettings.oauthUrl}/${USER_TOKEN_ENDPOINTS.EXCHANGE}?${q}`,
      params.exchangeRequest
    );
    return res.data;
  }
}

exports.USER_TOKEN_ENDPOINTS = USER_TOKEN_ENDPOINTS;
exports.UserTokenClient = UserTokenClient;
