// @microsoft/teams.api/dist/clients/bot/sign-in.js
'use strict';

var qs = require('qs');
var http = require('@microsoft/teams.common/http');
var apiClientSettings = require('../api-client-settings');

function _interopDefault (e) { return e && e.__esModule ? e : { default: e }; }

var qs__default = /*#__PURE__*/_interopDefault(qs);

const BOT_SIGNIN_ENDPOINTS = {
  URL: "api/botsignin/GetSignInUrl",
  RESOURCE: "api/botsignin/GetSignInResource"
};
class BotSignInClient {
  get http() {
    return this._http;
  }
  set http(v) {
    this._http = v;
  }
  _http;
  _apiClientSettings;
  constructor(options, apiClientSettings$1) {
    if (!options) {
      this._http = new http.Client();
    } else if ("request" in options) {
      this._http = options;
    } else {
      this._http = new http.Client(options);
    }
    this._apiClientSettings = apiClientSettings.mergeApiClientSettings(apiClientSettings$1);
  }
  async getUrl(params) {
    const q = qs__default.default.stringify(params);
    const res = await this.http.get(
      `${this._apiClientSettings.oauthUrl}/${BOT_SIGNIN_ENDPOINTS.URL}?${q}`
    );
    return res.data;
  }
  async getResource(params) {
    const q = qs__default.default.stringify(params);
    const res = await this.http.get(
      `${this._apiClientSettings.oauthUrl}/${BOT_SIGNIN_ENDPOINTS.RESOURCE}?${q}`
    );
    return res.data;
  }
}

exports.BOT_SIGNIN_ENDPOINTS = BOT_SIGNIN_ENDPOINTS;
exports.BotSignInClient = BotSignInClient;
