// @microsoft/teams.api/dist/clients/bot/sign-in.mjs
import qs from 'qs';
import { Client } from '@microsoft/teams.common/http';
import { mergeApiClientSettings } from '../api-client-settings';

const BOT_SIGNIN_ENDPOINTS = {
  URL: "api/botsignin/GetSignInUrl",
  RESOURCE: "api/botsignin/GetSignInResource"
};
class BotSignInClient {
  get http() {
    return this._http;
  }
  set http(v) {
    this._http = v;
  }
  _http;
  _apiClientSettings;
  constructor(options, apiClientSettings) {
    if (!options) {
      this._http = new Client();
    } else if ("request" in options) {
      this._http = options;
    } else {
      this._http = new Client(options);
    }
    this._apiClientSettings = mergeApiClientSettings(apiClientSettings);
  }
  async getUrl(params) {
    const q = qs.stringify(params);
    const res = await this.http.get(
      `${this._apiClientSettings.oauthUrl}/${BOT_SIGNIN_ENDPOINTS.URL}?${q}`
    );
    return res.data;
  }
  async getResource(params) {
    const q = qs.stringify(params);
    const res = await this.http.get(
      `${this._apiClientSettings.oauthUrl}/${BOT_SIGNIN_ENDPOINTS.RESOURCE}?${q}`
    );
    return res.data;
  }
}

export { BOT_SIGNIN_ENDPOINTS, BotSignInClient };
