// @microsoft/teams.apps/dist/middleware/auth/jwt-validator.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.createServiceTokenValidator = exports.createEntraTokenValidator = exports.JwtValidator = void 0;
const jsonwebtoken_1 = __importDefault(require("jsonwebtoken"));
const jwks_rsa_1 = __importDefault(require("jwks-rsa"));
const teams_common_1 = require("@microsoft/teams.common");
const asserts_1 = require("../../utils/asserts");
const DEFAULTS = {
    clockTolerance: 300 // 5 minutes
};
class JwtValidator {
    options;
    logger;
    jwksCache = new Map();
    constructor(options, logger) {
        this.options = options;
        this.logger = logger?.child('jwt-validator') ?? new teams_common_1.ConsoleLogger('jwt-validator');
    }
    /**
     * Validates a JWT token using the configured options
     */
    async validateAccessToken(rawToken, overrideOptions) {
        if (!rawToken) {
            throw new Error('No token provided');
        }
        return new Promise((resolve) => {
            const verifyOptions = {
                audience: [
                    this.options.clientId,
                    `api://${this.options.clientId}`,
                ],
                issuer: undefined,
                ignoreExpiration: false,
                algorithms: ['RS256'],
                clockTolerance: this.options.clockTolerance ?? DEFAULTS.clockTolerance
            };
            this.logger?.debug('Validating JWT token with options:', {
                audience: verifyOptions.audience,
                clockTolerance: verifyOptions.clockTolerance,
                algorithms: verifyOptions.algorithms
            });
            jsonwebtoken_1.default.verify(rawToken, this.getSigningKey.bind(this), verifyOptions, (err, decoded) => {
                if (err) {
                    this.logger?.error('JWT verification failed:', err);
                    resolve(null);
                    return;
                }
                if (!decoded || typeof decoded !== 'object') {
                    this.logger?.error('Decoded token is not a valid object:', decoded);
                    resolve(null);
                    return;
                }
                this.logger?.debug('JWT verification succeeded');
                const payload = decoded;
                try {
                    this.performCustomValidations(payload, overrideOptions);
                    this.logger?.debug('Custom validations passed for token');
                    resolve(payload);
                }
                catch (validationError) {
                    this.logger?.error('Custom validation failed:', validationError);
                    resolve(null);
                }
            });
        });
    }
    getJwksClient() {
        switch (this.options.jwksUriOptions.type) {
            case 'tenantId':
                {
                    const cachedClient = this.jwksCache.get(`${this.options.tenantId}`);
                    if (cachedClient) {
                        this.logger?.debug(`Using cached JWKS client for tenant ID: ${this.options.tenantId}`);
                        return cachedClient;
                    }
                    this.jwksCache.set(`${this.options.tenantId}`, (0, jwks_rsa_1.default)({
                        jwksUri: `https://login.microsoftonline.com/${this.options.tenantId}/discovery/v2.0/keys`,
                    }));
                    return this.jwksCache.get(`${this.options.tenantId}`);
                }
            case 'uri':
                {
                    const cachedClient = this.jwksCache.get(this.options.jwksUriOptions.uri);
                    if (cachedClient) {
                        this.logger?.debug(`Using cached JWKS client for URI: ${this.options.jwksUriOptions.uri}`);
                        return cachedClient;
                    }
                    this.jwksCache.set(this.options.jwksUriOptions.uri, (0, jwks_rsa_1.default)({
                        jwksUri: this.options.jwksUriOptions.uri,
                    }));
                    return this.jwksCache.get(this.options.jwksUriOptions.uri);
                }
            default:
                (0, asserts_1.assertNever)(this.options.jwksUriOptions, `Unknown JWKS URI options type: ${this.options.jwksUriOptions}`);
        }
    }
    getSigningKey(header, callback) {
        const jwksClient = this.getJwksClient();
        jwksClient?.getSigningKey(header.kid, (err, key) => {
            if (err) {
                this.logger?.error('Failed to get signing key:', err);
                callback(err, undefined);
                return;
            }
            const signingKey = key?.getPublicKey();
            callback(null, signingKey);
        });
    }
    validateIssuer(iss) {
        if (!this.options.validateIssuer) {
            return; // No issuer validation configured
        }
        if (!iss) {
            throw new Error('Token missing issuer claim');
        }
        if ('allowedIssuer' in this.options.validateIssuer) {
            // Validate against a specific allowed issuer
            if (iss !== this.options.validateIssuer.allowedIssuer) {
                throw new Error(`Token issuer '${iss}' does not match allowed issuer '${this.options.validateIssuer.allowedIssuer}'`);
            }
            return;
        }
        if (!this.options.tenantId) {
            return;
        }
        const isMultiTenant = ['common', 'organizations', 'consumers'].includes(this.options.tenantId);
        const allowedTenantIds = [];
        if (isMultiTenant) {
            if (this.options.validateIssuer.allowedTenantIds) {
                // find which tenant ids are not 'common', 'organizations', or 'consumers'
                for (const tenantId of this.options.validateIssuer.allowedTenantIds) {
                    if (!['common', 'organizations', 'consumers'].includes(tenantId)) {
                        allowedTenantIds.push(tenantId);
                    }
                }
            }
        }
        else {
            // For single-tenant apps, only allow tokens issued by this app's tenant
            // (ignore allowedTenantIds option for single-tenant apps)
            allowedTenantIds.push(this.options.tenantId);
        }
        if (allowedTenantIds.length === 0) {
            return; // No allowed tenant IDs configured, so no validation needed
        }
        else {
            // Validate against allowed tenant IDs
            if (!allowedTenantIds.some((tenantId) => iss.startsWith(`https://login.microsoftonline.com/${tenantId}/`))) {
                throw new Error(`Token issuer '${iss}' not in allowed tenant IDs: ${allowedTenantIds.join(', ')}`);
            }
        }
    }
    validateScope(scp, overrideValidateScope) {
        const validateScope = overrideValidateScope || this.options.validateScope;
        if (validateScope) {
            const scopes = scp ?? '';
            if (!scopes.includes(validateScope.requiredScope)) {
                throw new Error(`Token missing required scope: ${validateScope.requiredScope}`);
            }
        }
    }
    validateServiceUrl(serviceUrl, overrideValidateServiceUrl) {
        const validateServiceUrl = overrideValidateServiceUrl || this.options.validateServiceUrl;
        if (validateServiceUrl) {
            if (!serviceUrl) {
                throw new Error('Token missing serviceurl claim');
            }
            const normalizedTokenUrl = serviceUrl.replace(/\/$/, '').toLowerCase();
            const normalizedExpectedUrl = validateServiceUrl.expectedServiceUrl.replace(/\/$/, '').toLowerCase();
            if (normalizedTokenUrl !== normalizedExpectedUrl) {
                throw new Error(`Service URL mismatch. Token: ${normalizedTokenUrl}, Expected: ${normalizedExpectedUrl}`);
            }
        }
    }
    performCustomValidations(payload, overrideOptions) {
        this.validateIssuer(payload.iss);
        this.validateScope(payload.scp, overrideOptions?.validateScope);
        this.validateServiceUrl(payload.serviceurl, overrideOptions?.validateServiceUrl);
    }
}
exports.JwtValidator = JwtValidator;
// Factory functions for common scenarios
const createEntraTokenValidator = (tenantId, clientId, options) => {
    return new JwtValidator({
        clientId,
        tenantId,
        validateIssuer: {
            allowedTenantIds: options?.allowedTenantIds
        },
        validateScope: options?.requiredScope ? { requiredScope: options.requiredScope } : undefined,
        jwksUriOptions: {
            type: 'tenantId'
        },
    }, options?.logger);
};
exports.createEntraTokenValidator = createEntraTokenValidator;
const createServiceTokenValidator = (appId, tenantId, serviceUrl, logger) => {
    return new JwtValidator({
        clientId: appId,
        tenantId,
        validateIssuer: { allowedIssuer: 'https://api.botframework.com' },
        validateServiceUrl: serviceUrl ? { expectedServiceUrl: serviceUrl } : undefined,
        jwksUriOptions: {
            type: 'uri',
            uri: 'https://login.botframework.com/v1/.well-known/keys'
        },
    }, logger);
};
exports.createServiceTokenValidator = createServiceTokenValidator;
