// @microsoft/teams.apps/dist/app.oauth.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.onTokenExchange = onTokenExchange;
exports.onVerifyState = onVerifyState;
const axios_1 = require("axios");
const graph = __importStar(require("@microsoft/teams.graph"));
async function onTokenExchange(ctx) {
    const { api, activity, log, next } = ctx;
    if (this.oauth.defaultConnectionName !== activity.value.connectionName) {
        log.warn(`default connection name "${this.oauth.defaultConnectionName}" does not match activity connection name "${activity.value.connectionName}"`);
    }
    try {
        const token = await api.users.token.exchange({
            channelId: activity.channelId,
            userId: activity.from.id,
            connectionName: activity.value.connectionName,
            exchangeRequest: {
                token: activity.value.token,
            },
        });
        ctx.userGraph = new graph.Client(this.client.clone({
            token: token.token,
        }));
        this.events.emit('signin', { ...ctx, token, isSignedIn: true });
        next(ctx);
        return { status: 200 };
    }
    catch (error) {
        if (error instanceof axios_1.AxiosError) {
            if (error.status !== 404 && error.status !== 400 && error.status !== 412) {
                this.events.emit('error', { error, activity });
                return { status: error.status || 500 };
            }
        }
        return {
            status: 412,
            body: {
                id: activity.value.id,
                connectionName: activity.value.connectionName,
                failureDetail: 'unable to exchange token...',
            },
        };
    }
}
async function onVerifyState(ctx) {
    const { log, api, activity, next } = ctx;
    try {
        if (!activity.value.state) {
            log.warn(`auth state not found for conversation "${activity.conversation.id}" and user "${activity.from.id}"`);
            return { status: 404 };
        }
        const token = await api.users.token.get({
            channelId: activity.channelId,
            userId: activity.from.id,
            connectionName: this.oauth.defaultConnectionName,
            code: activity.value.state,
        });
        ctx.userGraph = new graph.Client(this.client.clone({
            token: token.token,
        }));
        this.events.emit('signin', { ...ctx, token, isSignedIn: true });
        next(ctx);
        return { status: 200 };
    }
    catch (error) {
        if (error instanceof axios_1.AxiosError) {
            if (error.status !== 404 && error.status !== 400 && error.status !== 412) {
                this.events.emit('error', { error, activity });
                return { status: error.status || 500 };
            }
        }
        return { status: 412 };
    }
}
