// @microsoft/teams.apps/dist/app.process.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.$process = $process;
const teams_api_1 = require("@microsoft/teams.api");
const api_1 = require("./api");
const contexts_1 = require("./contexts");
/**
 * activity handler called when an inbound activity is received
 * @param sender the plugin to use for sending activities
 * @param event the received activity event
 */
async function $process(sender, event) {
    const { token, activity } = event;
    this.log.debug(`activity/${activity.type}${activity.type === 'invoke' ? `/${activity.name}` : ''}`);
    let serviceUrl = activity.serviceUrl || token.serviceUrl;
    if (serviceUrl.endsWith('/')) {
        serviceUrl = serviceUrl.slice(0, serviceUrl.length - 1);
    }
    let userToken;
    try {
        userToken = await this.getUserToken(activity.channelId, activity.from.id);
    }
    catch (err) {
        // noop
    }
    const client = this.client.clone();
    const apiClient = new api_1.ApiClient(serviceUrl, this.client.clone({ token: () => this.getBotToken() }), this.options.apiClientSettings);
    const userGraph = new api_1.GraphClient(client.clone({ token: () => userToken }));
    const appGraph = new api_1.GraphClient(client.clone({ token: () => this.getAppGraphToken(activity.conversation.tenantId ?? 'common') }));
    const ref = {
        serviceUrl,
        activityId: activity.id,
        bot: activity.recipient,
        channelId: activity.channelId,
        conversation: activity.conversation,
        locale: activity.locale,
        user: activity.from,
    };
    const routes = this.router.select(activity);
    let pluginContexts = {};
    for (let i = this.plugins.length - 1; i > -1; i--) {
        const plugin = this.plugins[i];
        if (plugin.onActivity) {
            routes.unshift(async ({ next }) => {
                const additionalPluginContext = await plugin.onActivity({
                    ...ref,
                    sender: sender,
                    activity,
                    token,
                });
                if (additionalPluginContext) {
                    for (const key in additionalPluginContext) {
                        if (key in pluginContexts) {
                            this.log.warn(`Plugin context key "${key}" already exists. Overriding.`);
                        }
                    }
                    pluginContexts = {
                        ...pluginContexts,
                        ...additionalPluginContext,
                    };
                }
                return next();
            });
        }
    }
    let i = -1;
    let data = undefined;
    const next = async (ctx) => {
        if (i === routes.length - 1)
            return data;
        i++;
        const res = await routes[i](ctx || context.toInterface());
        if (res) {
            data = res;
        }
        return data;
    };
    const context = new contexts_1.ActivityContext(sender, {
        ...event,
        next,
        api: apiClient,
        userGraph,
        appGraph,
        appId: this.id || '',
        log: this.log,
        userToken: userToken,
        ref,
        storage: this.storage,
        isSignedIn: !!userToken,
        connectionName: this.oauth.defaultConnectionName,
        ...pluginContexts
    });
    const send = context.send.bind(context);
    context.send = async (activity, conversationRef) => {
        const res = await send(activity, conversationRef);
        this.onActivitySent(sender, {
            ...(conversationRef ?? ref),
            sender,
            activity: res,
        });
        return res;
    };
    context.stream.events.on('chunk', (activity) => {
        this.onActivitySent(sender, {
            ...ref,
            sender,
            activity,
        });
    });
    context.stream.events.once('close', (activity) => {
        this.onActivitySent(sender, {
            ...ref,
            sender,
            activity,
        });
    });
    let response;
    try {
        const res = await next();
        await context.stream.close();
        if ((0, teams_api_1.isInvokeResponse)(res)) {
            response = res;
        }
        else {
            response = { status: 200, body: res };
        }
        this.onActivityResponse(sender, {
            ...ref,
            sender,
            activity,
            response: res,
        });
    }
    catch (error) {
        response = { status: 500 };
        this.onError({ error, activity, sender });
        this.onActivityResponse(sender, {
            ...ref,
            sender,
            activity,
            response: response,
        });
    }
    return response;
}
