// @microsoft/teams.apps/dist/app.embed.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.func = func;
exports.tab = tab;
exports.configTab = configTab;
const path_1 = __importDefault(require("path"));
const middleware_1 = require("./middleware");
const utils_1 = require("./utils");
/**
 * add/update a function that can be called remotely
 * @param name The unique function name
 * @param cb The callback to handle the function
 */
function func(name, cb) {
    const log = this.log.child('functions').child(name);
    this.http.post(`/api/functions/${name}`, (0, middleware_1.withRemoteFunctionJwtValidation)({
        logger: log,
        entraTokenValidator: this.entraTokenValidator,
        ...this.credentials,
    }), async (req, res) => {
        if (!req.context) {
            throw new Error('expected client context');
        }
        const getCurrentConversationId = utils_1.functionContext.getConversationIdResolver(this, log.child('getCurrentConversationId'), req.context);
        const send = async (activity) => {
            const conversationId = await getCurrentConversationId();
            return !conversationId
                ? null
                : await this.send(conversationId, activity);
        };
        const data = await cb({
            ...req.context,
            log,
            api: this.api,
            appGraph: this.graph,
            data: req.body,
            getCurrentConversationId,
            send,
        });
        res.send(data);
    });
    return this;
}
/**
 * add/update a static tab.
 * the tab will be hosted at
 * `http://localhost:{{PORT}}/tabs/{{name}}` or `https://{{BOT_DOMAIN}}/tabs/{{name}}`
 * @remark scopes default to `personal`
 * @param name A unique identifier for the entity which the tab displays.
 * @param path The path to the web `dist` folder.
 */
function tab(name, path, options) {
    if (!this._manifest.staticTabs) {
        this._manifest.staticTabs = [];
    }
    const i = this._manifest.staticTabs.findIndex((t) => t.entityId === name);
    const tab = {
        entityId: name,
        contentUrl: `https://\${{BOT_DOMAIN}}/tabs/${name}`,
        scopes: ['personal'],
        ...options,
    };
    if (i > -1) {
        this._manifest.staticTabs[i] = tab;
    }
    else {
        this._manifest.staticTabs.push(tab);
    }
    this.http.static(`/tabs/${name}`, path);
    this.http.use(`/tabs/${name}*`, async (_, res) => {
        res.sendFile(path_1.default.join(path, 'index.html'));
    });
    return this;
}
/**
 * add a configurable tab
 * @remark scopes defaults to `team`
 * @param url The url to use when configuring the tab.
 */
function configTab(url, options) {
    if (!this._manifest.configurableTabs) {
        this._manifest.configurableTabs = [];
    }
    this._manifest.configurableTabs.push({
        configurationUrl: url,
        scopes: ['team'],
        ...options,
    });
    return this;
}
