// @microsoft/teams.apps/dist/token-manager.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.TokenManager = void 0;
const msal_node_1 = require("@azure/msal-node");
const teams_api_1 = require("@microsoft/teams.api");
const teams_common_1 = require("@microsoft/teams.common");
const DEFAULT_BOT_TOKEN_SCOPE = 'https://api.botframework.com/.default';
const DEFAULT_GRAPH_TOKEN_SCOPE = 'https://graph.microsoft.com/.default';
const DEFAULT_TENANT_FOR_BOT_TOKEN = 'botframework.com';
const DEFAULT_TENANT_FOR_GRAPH_TOKEN = 'common';
const GET_DEFAULT_TOKEN_AUTHORITY = (tenantId) => `https://login.microsoftonline.com/${tenantId}`;
const MSAL_LOG_LEVEL_TO_LOG_LEVEL = {
    [msal_node_1.LogLevel.Error]: 'error',
    [msal_node_1.LogLevel.Warning]: 'warn',
    [msal_node_1.LogLevel.Info]: 'info',
    [msal_node_1.LogLevel.Verbose]: 'debug',
    [msal_node_1.LogLevel.Trace]: 'trace'
};
const LOG_LEVEL_TO_MSAL_LOG_LEVEL = {
    'error': msal_node_1.LogLevel.Error,
    'warn': msal_node_1.LogLevel.Warning,
    'info': msal_node_1.LogLevel.Info,
    'debug': msal_node_1.LogLevel.Verbose,
    'trace': msal_node_1.LogLevel.Trace
};
// Type guard functions
function isClientCredentials(credentials) {
    return 'clientSecret' in credentials;
}
function isTokenCredentials(credentials) {
    return 'token' in credentials;
}
function isFederatedIdentityCredentials(credentials) {
    return 'managedIdentityType' in credentials;
}
class TokenManager {
    credentials;
    logger;
    _msalLogger;
    confidentialClientsByTenantId = {};
    managedIdentityClient = null;
    constructor(options, logger) {
        this.logger = logger.child('TokenManager') ?? new teams_common_1.ConsoleLogger('TokenManager');
        this._msalLogger = this.logger.child('azure/msal-node', {
            // Msal logging is fairly noisy. So we keep it quiet unless the user
            // explicitly turns it on
            pattern: '-azure/msal-node'
        });
        this.credentials = this.initializeCredentials(options);
    }
    async getBotToken() {
        return await this.getToken(DEFAULT_BOT_TOKEN_SCOPE, this.resolveTenantId(undefined, DEFAULT_TENANT_FOR_BOT_TOKEN));
    }
    async getGraphToken(tenantId) {
        return await this.getToken(DEFAULT_GRAPH_TOKEN_SCOPE, this.resolveTenantId(tenantId, DEFAULT_TENANT_FOR_GRAPH_TOKEN));
    }
    initializeCredentials(options) {
        const clientId = options.clientId ?? process.env.CLIENT_ID;
        const tenantId = options.tenantId ?? process.env.TENANT_ID;
        const clientSecret = options.clientSecret ?? process.env.CLIENT_SECRET;
        const token = options.token;
        const managedIdentityClientId = options.managedIdentityClientId ?? process.env.MANAGED_IDENTITY_CLIENT_ID;
        if (clientId && clientSecret) {
            this.logger.debug('Using Client Credentials auth');
            return {
                clientId,
                clientSecret,
                tenantId,
            };
        }
        else if (clientId && token) {
            this.logger.debug(('Using custom token factory auth'));
            return {
                clientId,
                tenantId,
                token,
            };
        }
        else if (clientId && !clientSecret) {
            if (managedIdentityClientId == null || managedIdentityClientId.toLowerCase() === clientId.toLowerCase()) {
                this.logger.debug('Using user managed identity auth');
                return {
                    clientId,
                    tenantId
                };
            }
            else {
                const identityType = managedIdentityClientId === 'system' ? 'system' : 'user';
                this.logger.debug(`Using Federated Identity Credentials auth (${identityType})`);
                return {
                    clientId,
                    tenantId,
                    managedIdentityClientId,
                    managedIdentityType: identityType,
                };
            }
        }
        return undefined;
    }
    async getToken(scope, tenantId) {
        if (!this.credentials) {
            return null;
        }
        if (isClientCredentials(this.credentials)) {
            return this.getTokenWithClientCredentials(this.credentials, scope, tenantId);
        }
        else if (isTokenCredentials(this.credentials)) {
            return this.getTokenWithTokenProvider(this.credentials, scope, tenantId);
        }
        else if (isFederatedIdentityCredentials(this.credentials)) {
            return this.getTokenWithFederatedCredentials(this.credentials, scope, tenantId);
        }
        else {
            return this.getTokenWithManagedIdentity(this.credentials, scope);
        }
    }
    async getTokenWithClientCredentials(credentials, scope, tenantId) {
        const confidentialClient = this.getConfidentialClient(credentials, tenantId);
        const result = await confidentialClient.acquireTokenByClientCredential({ scopes: [scope] });
        return this.handleTokenResponse(result);
    }
    async getTokenWithTokenProvider(credentials, scope, tenantId) {
        const token = await credentials.token(scope, tenantId);
        return new teams_api_1.JsonWebToken(token);
    }
    async getTokenWithManagedIdentity(credentials, scope) {
        const managedIdentityClient = this.getManagedIdentityClient(credentials);
        // Resource doesn't need the ./default suffix
        const resource = scope.replace('/.default', '');
        const result = await managedIdentityClient.acquireToken({
            resource
        });
        return this.handleTokenResponse(result);
    }
    async getTokenWithFederatedCredentials(credentials, scope, tenantId) {
        const managedIdentityClient = this.getManagedIdentityClient(credentials);
        const managedIdentityTokenRes = await managedIdentityClient.acquireToken({ resource: 'api://AzureADTokenExchange' });
        const confidentialClient = new msal_node_1.ConfidentialClientApplication({
            auth: {
                clientId: credentials.clientId,
                clientAssertion: managedIdentityTokenRes.accessToken,
                authority: GET_DEFAULT_TOKEN_AUTHORITY(tenantId)
            },
            system: {
                loggerOptions: this.buildLoggerOptions()
            }
        });
        const result = await confidentialClient.acquireTokenByClientCredential({ scopes: [scope] });
        return this.handleTokenResponse(result);
    }
    resolveTenantId(tenantId, defaultTenantId) {
        return tenantId || this.credentials?.tenantId || defaultTenantId;
    }
    getConfidentialClient(credentials, tenantId) {
        const cachedClient = this.confidentialClientsByTenantId[tenantId];
        if (cachedClient) {
            return cachedClient;
        }
        const client = new msal_node_1.ConfidentialClientApplication({
            auth: {
                clientId: credentials.clientId,
                clientSecret: credentials.clientSecret,
                authority: GET_DEFAULT_TOKEN_AUTHORITY(tenantId)
            },
            system: {
                loggerOptions: this.buildLoggerOptions()
            }
        });
        this.confidentialClientsByTenantId[tenantId] = client;
        return client;
    }
    getManagedIdentityClient(credentials) {
        if (this.managedIdentityClient) {
            return this.managedIdentityClient;
        }
        if (isFederatedIdentityCredentials(credentials)) {
            if (credentials.managedIdentityType === 'user') {
                this.managedIdentityClient = new msal_node_1.ManagedIdentityApplication({
                    managedIdentityIdParams: {
                        userAssignedClientId: credentials.managedIdentityClientId
                    },
                    system: {
                        loggerOptions: this.buildLoggerOptions()
                    }
                });
            }
            else {
                this.managedIdentityClient = new msal_node_1.ManagedIdentityApplication({
                    managedIdentityIdParams: undefined, //no options automatically indicates system assigned managed identity
                    system: {
                        loggerOptions: this.buildLoggerOptions()
                    }
                });
            }
        }
        else {
            this.managedIdentityClient = new msal_node_1.ManagedIdentityApplication({
                managedIdentityIdParams: {
                    userAssignedClientId: credentials.clientId
                },
                system: {
                    loggerOptions: this.buildLoggerOptions()
                }
            });
        }
        return this.managedIdentityClient;
    }
    handleTokenResponse(result) {
        if (!result) {
            throw new Error('Failed to get token');
        }
        return new teams_api_1.JsonWebToken(result.accessToken);
    }
    buildLoggerOptions() {
        return {
            logLevel: this.logger.loggerOptions?.level != null ? LOG_LEVEL_TO_MSAL_LOG_LEVEL[this.logger.loggerOptions.level] : undefined,
            loggerCallback: (level, message) => {
                // There's a bug in MSAL where it warns this on all requests.
                // https://github.com/AzureAD/microsoft-authentication-library-for-js/issues/7917
                if (message.endsWith('Warning - No client info in response')) {
                    return;
                }
                this._msalLogger.log(MSAL_LOG_LEVEL_TO_LOG_LEVEL[level], message);
            },
            piiLoggingEnabled: false,
        };
    }
}
exports.TokenManager = TokenManager;
