// @microsoft/teams.apps/dist/contexts/activity.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ActivityContext = void 0;
const teams_api_1 = require("@microsoft/teams.api");
class ActivityContext {
    appId;
    activity;
    ref;
    log;
    api;
    appGraph;
    userGraph;
    storage;
    stream;
    isSignedIn;
    connectionName;
    next;
    _plugin;
    _next;
    constructor(plugin, value) {
        Object.assign(this, value);
        this._plugin = plugin;
        this.stream = plugin.createStream(value.ref);
        this.connectionName = value.connectionName;
        if (value.activity.type === 'message') {
            value.activity = teams_api_1.MessageActivity.from(value.activity).toInterface();
        }
        if (value.activity.type === 'messageUpdate') {
            value.activity = teams_api_1.MessageUpdateActivity.from(value.activity).toInterface();
        }
        if (value.activity.type === 'messageDelete') {
            value.activity = teams_api_1.MessageDeleteActivity.from(value.activity).toInterface();
        }
        if (value.activity.type === 'typing') {
            value.activity = teams_api_1.TypingActivity.from(value.activity).toInterface();
        }
    }
    async send(activity, conversationRef) {
        return await this._plugin.send((0, teams_api_1.toActivityParams)(activity), conversationRef ?? this.ref);
    }
    async reply(activity) {
        activity = (0, teams_api_1.toActivityParams)(activity);
        activity.replyToId = this.activity.id;
        if (activity.type === 'message' && activity.text) {
            const blockQuote = this.buildBlockQuoteForActivity();
            if (blockQuote) {
                activity.text = `${blockQuote}\r\n${activity.text}`;
            }
        }
        return this.send(activity);
    }
    async signin(options) {
        const { oauthCardText, signInButtonText, connectionName, signInLink, overrideSignInActivity } = {
            oauthCardText: 'Please Sign In...',
            signInButtonText: 'Sign In',
            ...options,
        };
        const convo = { ...this.ref };
        try {
            const res = await this.api.users.token.get({
                channelId: this.activity.channelId,
                userId: this.activity.from.id,
                connectionName: connectionName || this.connectionName,
            });
            return res.token;
        }
        catch (err) {
            // noop
        }
        const tokenExchangeState = {
            connectionName: connectionName || this.connectionName,
            conversation: convo,
            relatesTo: this.activity.relatesTo,
            msAppId: this.appId,
        };
        if (this.activity.conversation.isGroup) {
            // create new 1:1 conversation with user to do SSO
            // because groupchats don't support it.
            const res = await this.api.conversations.create({
                tenantId: this.activity.conversation.tenantId,
                isGroup: false,
                bot: { id: this.activity.recipient.id },
                members: [this.activity.from],
            });
            await this.send({ type: 'message', text: oauthCardText });
            convo.conversation = { id: res.id };
        }
        const state = Buffer.from(JSON.stringify(tokenExchangeState)).toString('base64');
        const resource = await this.api.bots.signIn.getResource({ state });
        await this.send(overrideSignInActivity?.(resource.tokenExchangeResource, resource.tokenPostResource, resource.signInLink) ?? {
            type: 'message',
            inputHint: 'acceptingInput',
            recipient: this.activity.from,
            conversation: convo.conversation,
            attachments: [
                (0, teams_api_1.cardAttachment)('oauth', {
                    text: oauthCardText,
                    connectionName: connectionName || this.connectionName,
                    tokenExchangeResource: resource.tokenExchangeResource,
                    tokenPostResource: resource.tokenPostResource,
                    buttons: [
                        {
                            type: 'signin',
                            title: signInButtonText,
                            value: signInLink || resource.signInLink,
                        },
                    ],
                }),
            ],
        }, convo);
    }
    async signout(connectionName) {
        await this.api.users.token.signOut({
            channelId: this.activity.channelId,
            userId: this.activity.from.id,
            connectionName: connectionName || this.connectionName,
        });
    }
    toInterface() {
        return {
            activity: this.activity,
            api: this.api,
            appGraph: this.appGraph,
            userGraph: this.userGraph,
            appId: this.appId,
            log: this.log,
            ref: this.ref,
            storage: this.storage,
            stream: this.stream,
            isSignedIn: this.isSignedIn,
            connectionName: this.connectionName,
            userToken: this.userToken,
            next: this.next.bind(this),
            reply: this.reply.bind(this),
            send: this.send.bind(this),
            signin: this.signin.bind(this),
            signout: this.signout.bind(this),
        };
    }
    buildBlockQuoteForActivity() {
        if (this.activity.type === 'message' && this.activity.text) {
            const maxLength = 120;
            const truncatedText = this.activity.text.length > maxLength
                ? `${this.activity.text.substring(0, maxLength)}...`
                : this.activity.text;
            return `<blockquote itemscope="" itemtype="http://schema.skype.com/Reply" itemid="${this.activity.id}">
<strong itemprop="mri" itemid="${this.activity.from.id}">${this.activity.from.name}</strong><span itemprop="time" itemid="${this.activity.id}"></span>
<p itemprop="preview">${truncatedText}</p>
</blockquote>`;
        }
        else {
            this.log.debug('Skipping building blockquote for activity type:', this.activity.type);
        }
        return null;
    }
}
exports.ActivityContext = ActivityContext;
